src/agent/src/initdata.rs

```diff
@@ -0,0 +1,239 @@
+                debug!(
+                    logger,
+                    "Initdata read device `{filename}` first 4 bytes: {magic:?}"
+                );
+                if magic == SQUASHFS_MAGIC {
+                    let path = path.as_path().to_string_lossy().to_string();
+                    debug!(logger, "Found candidate SquashFS device: {path}");
+                    squashfs_devices.push(path);
+                }
+            }
+            Err(e) => debug!(logger, "Initdata read device `{filename}` failed: {e:?}"),
+        }
+    }
+
+    Ok(squashfs_devices)
+}
+
+pub fn try_read_initdata(device_path: &str) -> Result<Option<Vec<u8>>> {
+    let initdata_devfile = std::fs::File::open(device_path)?;
+    let buf_reader = BufReader::new(initdata_devfile);
+    let read_filesystem = FilesystemReader::from_reader(buf_reader)?;
+
+    let mut initdata_content = Vec::new();
+
+    // initdata squash device only allow 1 file `initdata.toml` inside
+    for node in read_filesystem.files() {
+        // skip root
+        if node.fullpath.as_path() == Path::new("/") {
+            continue;
+        }
+
+        if node.fullpath.as_path() != Path::new(INITDATA_FILE_IN_SQUASHFS) {
+            bail!(
+                "Not a legal initdata device! get path {:?}",
+                node.fullpath.as_path()
+            );
+        }
+
+        let InnerNode::File(file) = &node.inner else {
+            bail!("Not a legal initdata device! {INITDATA_FILE_IN_SQUASHFS} is not a file!");
+        };
+
+        let mut reader = read_filesystem.file(&file).reader();
+        let n = reader.read_to_end(&mut initdata_content)?;
+        if n > 0 {
+            return Ok(Some(initdata_content));
+        }
+        break;
+    }
+
+    Ok(None)
+}
+
+pub async fn initialize_initdata(logger: &Logger) -> Result<()> {
+    let logger = logger.new(o!("subsystem" => "initdata"));
+    let candidate_initdata_devices = detect_candidate_initdata_devices(&logger).await?;
+    if candidate_initdata_devices.is_empty() {
+        info!(
+            logger,
+            "Initdata device not found, skip initdata initialization"
+        );
+        return Ok(());
+    };
+
+    tokio::fs::create_dir_all(INITDATA_PATH)
+        .await
+        .inspect_err(|e| error!(logger, "Failed to create initdata dir: {e:?}"))?;
+
+    let initdata_plaintext_option = candidate_initdata_devices
+        .iter()
+        .filter_map(|device_path| match try_read_initdata(device_path) {
+            Ok(r) => r,
+            Err(e) => {
+                debug!(
+                    logger,
+                    "Failed to read initdata from device {device_path}: {e:?}"
+                );
+                None
+            }
+        })
+        .next();
+
+    let Some(initdata_content) = initdata_plaintext_option else {
+        info!(
+            logger,
+            "Initdata device not found, skip initdata initialization"
+        );
+        return Ok(());
+    };
```

Comment:
are we being too graceful here? If we found an initdata block device, initdata should be present, no?

should we accept multiple device candidates or fail, if we detect more then one?

Reply:
This is to avoid the issue mentioned in https://github.com/kata-containers/kata-containers/pull/10610#discussion_r1983365666, if there is a pod using squashfs rbd. So here we try to find a special rbd with the file "/.kata.initdata.toml" - we believe this path is special enough that hard to meet collisionwill.